crypto/asn1/tasn_dec.c

```diff
@@ -921,6 +921,14 @@ static int asn1_ex_c2i(ASN1_VALUE **pval, const unsigned char *cont, int len,
             ERR_raise(ERR_LIB_ASN1, ASN1_R_UNIVERSALSTRING_IS_WRONG_LENGTH);
             goto err;
         }
+        if (utype == V_ASN1_GENERALIZEDTIME && (len < 15)) {
+            ERR_raise(ERR_LIB_ASN1, ASN1_R_GENERALIZEDTIME_IS_TOO_SHORT);
```

Comment:
This check is more strict than what we have here:
https://github.com/openssl/openssl/blob/22f82d457c06289ec66a627a3d11649d83beff88/crypto/asn1/a_time.c#L102-L110
so I think, it is likely to cause issues.

Reply:
Can you substantiate what issues you expect to be caused? I'm skeptical implementers would consciously rely on invalid encodings, X.690 Section 11.7 reads unambiguous to me.

I'm not arguing for full conformance to RFC 5280 (although such strictness seems to work well in other crypto libraries), but to ensure that at the very least an application can't mistakenly interpret an invalid _GeneralizedTime_ as a valid _UTCTime_ (for example, if the decision tree is based on the length).

The code path followed in `a_time.c` when `strict` is set to `1` arguably is far stricter than what I propose here.

Reply:
I dont know why this check is more permissive, but your test vectors are rejected by ASN1_TIME_check,
but slightly modified test vectors i.e. without seconds pass the check in ASN1_TIME_check, as this experiment
shows:
```
diff --git a/test/asn1_decode_test.c b/test/asn1_decode_test.c
index f112dd7034..e33f946520 100644
--- a/test/asn1_decode_test.c
+++ b/test/asn1_decode_test.c
@@ -166,10 +166,10 @@ static int test_uint64(void)
 
 static int test_gentime(void)
 {
-    /* Underflowing GeneralizedTime 161208193400Z (YYMMDDHHMMSSZ) */
+    /* Underflowing GeneralizedTime 209912312359Z (YYYYMMDDHHMMZ) */
     const unsigned char der[] = {
-        0x18, 0x0d, 0x31, 0x36, 0x31, 0x32, 0x30, 0x38, 0x31,
-        0x39, 0x33, 0x34, 0x30, 0x30, 0x5a,
+        0x18, 0x0d, 0x32, 0x30, 0x39, 0x39, 0x31, 0x32, 0x33,
+        0x31, 0x32, 0x33, 0x35, 0x39, 0x5a,
     };
     const unsigned char *p;
     int der_len, rc = 1;
@@ -179,7 +179,8 @@ static int test_gentime(void)
     der_len = sizeof(der);
     gentime = d2i_ASN1_GENERALIZEDTIME(NULL, &p, der_len);
 
-    if (!TEST_ptr_null(gentime))
+    if (!TEST_true(gentime != NULL
+            && ASN1_GENERALIZEDTIME_check(gentime)))
         rc = 0; /* fail */
 
     ASN1_GENERALIZEDTIME_free(gentime);
@@ -190,10 +191,10 @@ static int test_gentime(void)
 
 static int test_utctime(void)
 {
-    /* Underflowing UTCTime 0205104700Z (MMDDHHMMSSZ) */
+    /* Underflowing UTCTime 9912312359Z (YYMMDDHHMMZ) */
     const unsigned char der[] = {
-        0x17, 0x0b, 0x30, 0x32, 0x30, 0x35, 0x31, 0x30,
-        0x34, 0x37, 0x30, 0x30, 0x5a,
+        0x17, 0x0b, 0x39, 0x39, 0x31, 0x32, 0x33, 0x31,
+        0x32, 0x33, 0x35, 0x39, 0x5a,
     };
     const unsigned char *p;
     int der_len, rc = 1;
@@ -203,7 +204,8 @@ static int test_utctime(void)
     der_len = sizeof(der);
     utctime = d2i_ASN1_UTCTIME(NULL, &p, der_len);
 
-    if (!TEST_ptr_null(utctime))
+    if (!TEST_true(utctime != NULL
+            && ASN1_UTCTIME_check(utctime)))
         rc = 0; /* fail */
 
     ASN1_UTCTIME_free(utctime);

```

Reply:
so IMHO ASN1_XXXTIME_check should also be changed to not allow this any more,
and of course a CHANGES entry should be added documenting this change.

Reply:
Yes, `ASN1_TIME_check()` ultimately wraps around `ossl_asn1_utctime_to_tm()` which indeed does a few useful sanity checks, but on the `d2i_ASN1_GENERALIZEDTIME() / d2i_ASN1_UTCTIME()` paths, `ossl_asn1_utctime_to_tm()` isn't invoked.

Reply:
@bernd-edlinger I updated `ASN1_XXXTIME_check()` too